examples/matmul_layernorm.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import annotations
+
+import torch
+import torch.nn.functional as F
+
+import helion
+import helion.language as hl
+
+
+# static_shapes=True gives a performance boost for matmuls
+@helion.kernel(static_shapes=True)
+def matmul_layernorm(
+    x: torch.Tensor, y: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor
+) -> torch.Tensor:
+    m, k = x.size()
+    k2, n = y.size()
+    assert k == k2, f"size mismatch {k} != {k2}"
+    assert weight.size(0) == n, f"weight size mismatch {weight.size(0)} != {n}"
+    assert bias.size(0) == n, f"bias size mismatch {bias.size(0)} != {n}"
+    out = torch.empty(
+        [m, n], dtype=torch.promote_types(x.dtype, y.dtype), device=x.device
+    )
+    # NOTE: block_size=n for the second dimension is required for correctness
+    # since layernorm computation needs the entire row.
+    for tile_m, tile_n in hl.tile([m, n], block_size=[None, n]):
+        acc = hl.zeros([tile_m, tile_n], dtype=torch.float32)
+        for tile_k in hl.tile(k):
+            mm = torch.matmul(x[tile_m, tile_k], y[tile_k, tile_n])
+            acc = acc + mm
+        acc = F.layer_norm(
+            acc,
+            [acc.size(1)],
+            weight[tile_n].to(torch.float32),
+            bias[tile_n].to(torch.float32),
+        )
+        out[tile_m, tile_n] = acc
+    return out
```

Comment:
To avoid the UNPADDED_* stuff, you might be able to do:
```suggestion
    hl.specialize(n)
    for tile_m in hl.tile(m):
        acc = hl.zeros([tile_m, helion.next_power_of_two(n)], dtype=torch.float32)
        for tile_k in hl.tile(k):
            mm = torch.matmul(x[tile_m, tile_k], y[tile_k, :])
            acc = acc + mm
        acc = F.layer_norm(
            acc,
            [acc.size(1)],
            weight[tile_n].to(torch.float32),
            bias[tile_n].to(torch.float32),
        )
        out[tile_m, :] = acc
    return out
```

Reply:
We should be able to get:
```py
        acc[tile_m, :] = F.layer_norm(
            acc[tile_m, :],
            [n],
            weight.to(torch.float32),
            bias.to(torch.float32),
        )
```
to work.

There is already very similar code to the `_UNPADDED_SIZE_` thing for reduction dimensions.  If you do:
```py
x[:].mean()
```
This decomps to:
```py
x[:].sum() / x.size(0)
```
but we will use the original size of `x` rather than the padded one, because of:
https://github.com/pytorch-labs/helion/blob/c674b43dbee44b61034ed23041c27c9b94d73010/helion/_compiler/reduction_strategy.py#L56-L57

For reductions we fake having a non-power-of-2 block size.

Reply:
Curious if we should worry that when the user reads the code `acc[tile_m, :]` they would think that the 2nd dim is `0 ... next_power_of_two(n)` instead of `0 ... n`, since 2nd dim of `acc` was indeed defined as having `next_power_of_two(n)` as upper bound.

In that case, I feel that there is a chance they could get confused about why `acc` dim size is `next_power_of_two(n)` but we call `F.layer_norm(..., [n], ...)`.

Wonder would it be a good idea to always do the `[:n]` slicing, e.g. something like:
```python
        acc[tile_m, :n] = F.layer_norm(
            acc[tile_m, :n],
            [n],
            weight.to(torch.float32),
            bias.to(torch.float32),
        )
```
to make sure the dim sizes always appear aligned (i.e. always `n`)?

Reply:
I tried going down this path, but ran into an issue of `next_power_of_2(n)` and `n` being hard to reconcile together: it happens because the 2nd dim of `acc` is of `next_power_of_2(n)` size (explicitly set as specified in the kernel), but then we pass in `[n]` for the second arg of layer_norm() call. And then these two would create a conflict in shape checking during layer_norm lowering.

I tried the original suggestion of making `hl.specialize(n)` + `block_size=[n]` to work without having `_UNPADDED_SIZE`, and I believe I am able to get it to work without needing an extra `_UNPADDED_SIZE` arg. (Updated the PR)